feat: auto-plan projects when modules change

iamnoah opened the pull request:
Fixes #920

See [testdata](https://github.com/runatlantis/atlantis/pull/2507/files#diff-f2a614e806f20051c955922d44c5602829e844794bfcd938e5d91d35cdfae172) for an example repo structure. Any project that uses modules in the same repo should be able to take advantage. 

Only files that are not considered part of a project will be considered. With the current auto-plan heuristic, only changed files in `/modules/` without a `main.tf` would trigger a check of the modules dependencies. 

Adds `--autoplan-modules` and `--autoplan-modules-from-projects`

jamengual commented:
What are the repo structure and the module instantiation used?

a more detailed example could help to understand the feature

iamnoah commented:
@jamengual updated the description to point to the test data. It's layout agnostic as it parses the module dependencies out of the projects.

jamengual commented:
so this only works when the `source =` is on the local repo not when the `source = cloudposse/vpc/aws` which is the terraform registry ?

iamnoah commented:
@jamengual correct, it's only detecting changes to local modules.

iamnoah commented:
Sorry for the churn, in testing this out, I've squashed some bugs and cleaned things up. 

I think the docs now lay out the use case. Since it's a departure from current behavior, it requires a flag to activate, but I think `--autoplan-modules` does what one would intuitively expect. If the module indexing does something undesired, e.g., take a long time or fail on unusual projects, it can be disabled selectively by changing its pattern.

iamnoah commented:
Anything else I need to do to have this considered? A week of using it in our pipelines hasn't revealed any new issues.

pauloconnor commented:
This sounds similar to something I have in my `atlantis.yaml` generation script. I'm curious to know if you can disable autoplanning for individual child modules. For example, I have a child module that is called hundreds of times that I don't want to autoplan for obvious reasons, so I include a header in my `main.tf` file so my script knows to ignore it.

iamnoah commented:
@pauloconnor yup! `--autoplan-file-list=**/*.tf...,!modules/really-common-child-module/*` would keep changes to that module from being planned.

pauloconnor commented:
@iamnoah Damn, that's fantastic! Hopefully this'll get merged in soon.

nitrocode commented on runatlantis.io/docs/autoplanning.md:
```diff
@@ -8,7 +8,7 @@ The algorithm it uses is as follows:
 1. If the directory path doesn't contain `modules/` then try to run `plan` in that directory
 1. If it does contain `modules/` look at the directory one level above `modules/`. If it
-contains a `main.tf` run plan in that directory, otherwise ignore the change.
+contains a `main.tf` run plan in that directory, otherwise ignore the change (see below for exceptions.)
```
```suggestion
contains a `main.tf` run plan in that directory, otherwise ignore the change (see below for exceptions).
```

nitrocode commented on runatlantis.io/docs/server-configuration.md:
````diff
@@ -134,6 +134,53 @@ Values are chosen in this order:
+### `--autoplan-modules`
+
+```bash
+  atlantis server --autoplan-modules=true'
````
```suggestion
atlantis server --autoplan-modules=true'
```

nitrocode commented on runatlantis.io/docs/server-configuration.md:
````diff
@@ -134,6 +134,53 @@ Values are chosen in this order:
+### `--autoplan-modules-from-projects`
+
+```bash
+  atlantis server --autoplan-modules-from-projects='**/init.tf'
````
```suggestion
atlantis server --autoplan-modules-from-projects='**/init.tf'
```

nitrocode commented on server/events/project_finder_test.go:
```diff
@@ -296,7 +296,7 @@ func TestDetermineProjects(t *testing.T) {
 	for _, c := range cases {
 		t.Run(c.description, func(t *testing.T) {
-			projects := m.DetermineProjects(noopLogger, c.files, modifiedRepo, c.repoDir, c.autoplanFileList)
+			projects := m.DetermineProjects(noopLogger, c.files, modifiedRepo, c.repoDir, c.autoplanFileList, nil)
```
Do we need another test here?

iamnoah replied:
I don't think so? 

Relevant DetermineProjects code:

```go
		if moduleInfo != nil {
			downstreamProjects := moduleInfo.DependentProjects(path.Dir(modifiedFile))
			log.Debug("found downstream projects for %q: %v", modifiedFile, downstreamProjects)
			dirs = append(dirs, downstreamProjects...)
		}
```

and moduleInfo.DependentProjects is tested.

nitrocode commented on server/events/modules.go:
```diff
@@ -0,0 +1,161 @@
+
+type ModuleProjects interface {
+	// DownstreamProjects returns all projects that depend on the module at moduleDir
+	DownstreamProjects(moduleDir string) []string
```
A more accurate name could perhaps be DependentProjects? What do you think?

nitrocode commented:
@iamnoah please merge from the default branch to fix the `branch` test failure

jamengual commented:
@iamnoah can you fix the conflicts?

nitrocode commented:
@iamnoah please fix conflicts

nitrocode commented:
@iamnoah please fix linting error in failed branch pr check

nitrocode commented:
@iamnoah apologies but looks like this PR has some conflicts, could you resolve them?

jamengual commented:
@iamnoah  Thanks for the contribution

lattwood commented:
@jamengual what are the chances of getting a release cut with this in it? Looks like the merge happened less than two hours after the release was tagged.

jamengual commented:
we do pre-release for new features before they go to the stable branch.

We will do a prerelease soon.
